persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,680 @@
+        , quotation
+        , parenthetical
+        , ptext
+        ]
+
+class Block a where
+    blockFirstPos :: a -> SourcePos
+    blockMembers :: a -> [Member]
+    blockSetMembers :: [Member] -> a -> a
+    blockSetNELMembers :: NonEmpty Member -> a -> a
+    blockSetDocCommentBlock :: Maybe DocCommentBlock -> a -> a
+
+instance Block EntityBlock where
+    blockFirstPos = entityHeaderPos . entityBlockEntityHeader
+    blockMembers = entityBlockMembers
+    blockSetMembers ms b = b{entityBlockMembers = ms}
+    blockSetNELMembers nel = blockSetMembers (NEL.toList nel)
+    blockSetDocCommentBlock dcb b = b{entityBlockDocCommentBlock = dcb}
+
+instance Block ExtraBlock where
+    blockFirstPos = extraBlockHeaderPos . extraBlockExtraBlockHeader
+    blockMembers = NEL.toList . extraBlockMembers
+    blockSetMembers ms b = case NEL.nonEmpty ms of
+        Nothing -> b
+        Just nel -> blockSetNELMembers nel b
+    blockSetNELMembers nel b = b{extraBlockMembers = nel}
+    blockSetDocCommentBlock dcb b = b{extraBlockDocCommentBlock = dcb}
+
+blockLastPos :: (Block a) => a -> SourcePos
+blockLastPos b = case blockMembers b of
+    [] -> blockFirstPos b
+    members -> maximum $ fmap memberEndPos members
+
+blockBlockAttrs :: (Block a) => a -> [BlockAttr]
+blockBlockAttrs eb =
+    foldMap f (blockMembers eb)
+  where
+    f = \case
+        MemberBlockAttr fs -> [fs]
+        _ -> []
+
+blockExtraBlocks :: (Block a) => a -> [ExtraBlock]
+blockExtraBlocks eb =
+    foldMap f (blockMembers eb)
+  where
+    f = \case
+        MemberExtraBlock ex -> [ex]
+        _ -> []
+
+data ParsedEntityDef = ParsedEntityDef
+    { parsedEntityDefComments :: [Text]
+    , parsedEntityDefEntityName :: EntityNameHS
+    , parsedEntityDefIsSum :: Bool
+    , parsedEntityDefEntityAttributes :: [Attr]
+    , parsedEntityDefFieldAttributes :: [([Token], Maybe Text)]
+    , parsedEntityDefExtras :: M.Map Text [ExtraLine]
+    , parsedEntityDefSpan :: Maybe SourceSpan
+    }
+    deriving (Show)
+
+data DocCommentBlock = DocCommentBlock
+    { docCommentBlockLines :: [Text]
+    , docCommentBlockPos :: SourcePos
+    }
+    deriving (Show)
+
+data EntityHeader = EntityHeader
+    { entityHeaderSum :: Bool
+    , entityHeaderTableName :: Text
+    , entityHeaderRemainingTokens :: [Token]
+    , entityHeaderPos :: SourcePos
+    }
+    deriving (Show)
+
+data EntityBlock = EntityBlock
+    { entityBlockDocCommentBlock :: Maybe DocCommentBlock
+    , entityBlockEntityHeader :: EntityHeader
+    , entityBlockMembers :: [Member]
+    }
+    deriving (Show)
+
+data ExtraBlockHeader = ExtraBlockHeader
+    { extraBlockHeaderKey :: Text
+    , extraBlockHeaderRemainingTokens :: [Token]
+    , extraBlockHeaderPos :: SourcePos
+    }
+    deriving (Show)
+
+data ExtraBlock = ExtraBlock
+    { extraBlockDocCommentBlock :: Maybe DocCommentBlock
+    , extraBlockExtraBlockHeader :: ExtraBlockHeader
+    , extraBlockMembers :: NonEmpty Member
+    }
+    deriving (Show)
```

Comment:
So if we have `data Block` as mentioned above, then we can factor out the structure here:

```suggestion
data EntityBlock = EntityBlock
    { entityBlockBlock :: Block
    , entityBlockEntityHeader :: EntityHeader
    }
    deriving (Show)

data ExtraBlockHeader = ExtraBlockHeader
    { extraBlockHeaderKey :: Text
    , extraBlockHeaderRemainingTokens :: [Token]
    , extraBlockHeaderPos :: SourcePos
    }
    deriving (Show)

data ExtraBlock = ExtraBlock
    { extraBlockExtraBlockHeader :: ExtraBlockHeader
    , extraBlockBlock :: Block
    }
    deriving (Show)
```